avocado/core/job.py

```diff
@@ -564,6 +570,9 @@ def run(self):
             self.log.error('Report bugs visiting %s', _NEW_ISSUE_LINK)
             return exit_codes.AVOCADO_FAIL
         finally:
+            for ext in self.job_pre_post_dispatcher.extensions:
+                if hasattr(ext.obj, 'post'):
+                    ext.obj.post(self)
```

Comment:
I'm sorry, I fail to understand it completely. If I use your version and intentionally break the API:

```diff
diff --git a/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py b/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
index 3899ccb..48251f5 100644
--- a/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
+++ b/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
@@ -17,7 +17,7 @@ class Sleep(JobPre, JobPost):
                                           key_type=int,
                                           default=3)
 
-    def sleep(self, job):
+    def sleep(self, job, aaa):
         for i in xrange(1, self.seconds + 1):
             self.log.info("Sleeping %2i/%s", i, self.seconds)
             time.sleep(1)
```

Then it fails:

```
avocado run passtest
Avocado crashed:
Traceback (most recent call last):
  File "/bin/avocado", line 83, in <module>
    sys.exit(app.run())
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/core/app.py", line 69, in run
    return method(self.parser.args)
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/plugins/run.py", line 202, in run
    return job_instance.run()
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/core/job.py", line 547, in run
    ext.obj.pre(self)
TypeError: sleep() takes exactly 3 arguments (2 given)

Avocado crashed unexpectedly: sleep() takes exactly 3 arguments (2 given)
You can find details in /var/tmp/avocado-traceback-2016-04-20_08:20:37-RIVf7Z.log
```

When I remove all JobPre and JobPost:

```diff
diff --git a/avocado/plugins/base.py b/avocado/plugins/base.py
index 1cefa79..03337b7 100644
--- a/avocado/plugins/base.py
+++ b/avocado/plugins/base.py
@@ -84,35 +84,3 @@ class CLICmd(Plugin):
         """
         Entry point for actually running the command
         """
-
-
-class JobPre(Plugin):
-
-    """
-    Base plugin interface for adding actions before a job runs
-
-    Plugins that want to add actions to be run before a job runs,
-    should use the 'avocado.plugins.job.pre' namespace.
-    """
-
-    @abc.abstractmethod
-    def pre(self, job):
-        """
-        Entry point for actually running the pre job action
-        """
-
-
-class JobPost(Plugin):
-
-    """
-    Base plugin interface for adding actions after a job runs
-
-    Plugins that want to add actions to be run after a job runs,
-    should use the 'avocado.plugins.job.post' namespace.
-    """
-
-    @abc.abstractmethod
-    def post(self, job):
-        """
-        Entry point for actually running the post job action
-        """
diff --git a/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py b/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
index 3899ccb..81f9192 100644
--- a/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
+++ b/examples/plugins/job-pre-post/sleep/avocado_job_sleep.py
@@ -2,10 +2,9 @@ import time
 import logging
 
 from avocado.core.settings import settings
-from avocado.plugins.base import JobPre, JobPost
 
 
-class Sleep(JobPre, JobPost):
+class Sleep(object):
 
     name = 'sleep'
     description = 'Sleeps for a number of seconds'
@@ -17,7 +16,7 @@ class Sleep(JobPre, JobPost):
                                           key_type=int,
                                           default=3)
 
-    def sleep(self, job):
+    def sleep(self, job, aaa):
         for i in xrange(1, self.seconds + 1):
             self.log.info("Sleeping %2i/%s", i, self.seconds)
             time.sleep(1)
```

It crashes in the same way:

```
avocado run passtest
Avocado crashed:
Traceback (most recent call last):
  File "/bin/avocado", line 83, in <module>
    sys.exit(app.run())
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/core/app.py", line 69, in run
    return method(self.parser.args)
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/plugins/run.py", line 202, in run
    return job_instance.run()
  File "/home/medic/Work/Projekty/avocado/avocado/avocado/core/job.py", line 547, in run
    ext.obj.pre(self)
TypeError: sleep() takes exactly 3 arguments (2 given)

Avocado crashed unexpectedly: sleep() takes exactly 3 arguments (2 given)
You can find details in /var/tmp/avocado-traceback-2016-04-20_08:23:54-WAe2Iv.log
```

When I fix it by accepting only a single parameter, it works fine.

So the only difference is, that when I use `object` as a base class, it accepts anything and if `pre` or `post` exists, it tires to execute them and fails on incorrect interface (additional argumnets).

If I use `JobPre`, it fails additionally if I forget to implement `pre`, but it still crashes on an incorrect interface. The same `JobPost`.

So in the end the only difference is, that when the user uses `JobPre`/`JobPost` (which is optional), he is notified if he forgets about the related method, but he is free to use the whole interface:

```
# Using JobPre (no pre/post methods)
avocado run passtest
Failed to load plugin from module "avocado_job_sleep": TypeError("Can't instantiate abstract class Sleep with abstract methods pre",)
JOB ID     : 440c4a7facece8fa3d9cf55c022cfc45e2219693
JOB LOG    : /home/medic/avocado/job-results/job-2016-04-20T08.28-440c4a7/job.log
TESTS      : 1
 (1/1) passtest.py:PassTest.test: PASS
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /home/medic/avocado/job-results/job-2016-04-20T08.28-440c4a7/html/results.html
TIME       : 0.00 s

# Using JobPre (pre+post methods)
avocado run passtest
Sleeping  1/3
Sleeping  2/3
Sleeping  3/3
JOB ID     : 554d2721cc7c49c75a201e210ef97d0fd41f0275
JOB LOG    : /home/medic/avocado/job-results/job-2016-04-20T08.25-554d272/job.log
TESTS      : 1
 (1/1) passtest.py:PassTest.test: PASS
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /home/medic/avocado/job-results/job-2016-04-20T08.25-554d272/html/results.html
TIME       : 0.00 s
Sleeping  1/3
Sleeping  2/3
Sleeping  3/3
```
```
# Using object (no pre/post methods)
avocado run passtest
JOB ID     : fc69b64986309f09d53dbe4bf66a770efc160ba4
JOB LOG    : /home/medic/avocado/job-results/job-2016-04-20T08.26-fc69b64/job.log
TESTS      : 1
 (1/1) passtest.py:PassTest.test: PASS
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /home/medic/avocado/job-results/job-2016-04-20T08.26-fc69b64/html/results.html
TIME       : 0.00 s
# Using object (pre+post methods)
avocado run passtest
Sleeping  1/3
Sleeping  2/3
Sleeping  3/3
JOB ID     : 554d2721cc7c49c75a201e210ef97d0fd41f0275
JOB LOG    : /home/medic/avocado/job-results/job-2016-04-20T08.25-554d272/job.log
TESTS      : 1
 (1/1) passtest.py:PassTest.test: PASS
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /home/medic/avocado/job-results/job-2016-04-20T08.25-554d272/html/results.html
TIME       : 0.00 s
Sleeping  1/3
Sleeping  2/3
Sleeping  3/3
```